File: Python/count-number-of-ways-to-place-houses.py
# Time:  O(logn)
# Space: O(1)

import itertools


# matrix exponentiation
class Solution(object):
    def countHousePlacements(self, n):
        """
        :type n: int
        :rtype: int
        """
        MOD = 10**9+7
        def matrix_mult(A, B):
            ZB = zip(*B)
            return [[sum(a*b % MOD for a, b in itertools.izip(row, col)) % MOD for col in ZB] for row in A]
 
        def matrix_expo(A, K):
            result = [[int(i == j) for j in xrange(len(A))] for i in xrange(len(A))]
            while K:
                if K % 2:
                    result = matrix_mult(result, A)
                A = matrix_mult(A, A)
                K /= 2
            return result

        T = [[1, 1],
             [1, 0]]
        return pow(matrix_mult([[1,  0]], matrix_expo(T, n+1))[0][0], 2, MOD)  # [a1, a0] * T^N

    
# Time:  O(n)
# Space: O(1)
# dp
class Solution2(object):
    def countHousePlacements(self, n):
        """
        :type n: int
        :rtype: int
        """
        MOD = 10**9+7
        prev, curr = 0, 1
        for _ in xrange(n+1):
            prev, curr = curr, (prev+curr)%MOD
        return pow(curr, 2, MOD)

File: C++/count-number-of-ways-to-place-houses.cpp
// Time:  O(logn)
// Space: O(1)

// matrix exponentiation
class Solution {
public:
    int countHousePlacements(int n) {
        vector<vector<int>> T = {{1, 1},
                                 {1, 0}};
        int64_t result = matrixMult({{1, 0}}, matrixExpo(T, n + 1))[0][0]; // [a1, a0] * T^N
        return result * result % MOD;
    }

private:
    vector<vector<int>> matrixExpo(const vector<vector<int>>& A, int pow) {
        vector<vector<int>> result(A.size(), vector<int>(A.size()));
        vector<vector<int>> A_exp(A);
        for (int i = 0; i < A.size(); ++i) {
            result[i][i] = 1;
        }
        while (pow) {
            if (pow % 2 == 1) {
                result = matrixMult(result, A_exp);
            }
            A_exp = matrixMult(A_exp, A_exp);
            pow /= 2;
        }
        return result;
    }

    vector<vector<int>> matrixMult(const vector<vector<int>>& A, const vector<vector<int>>& B) {
        vector<vector<int>> result(A.size(), vector<int>(B[0].size()));
        for (int i = 0; i < A.size(); ++i) {
            for (int j = 0; j < B[0].size(); ++j) {
                int64_t entry = 0;
                for (int k = 0; k < B.size(); ++k) {
                    entry = (static_cast<int64_t>(A[i][k]) * B[k][j] % MOD + entry) % MOD;
                }
                result[i][j] = static_cast<int>(entry);
            }
        }
        return result;
    }
    const int MOD = 1e9 + 7;
};


// Time:  O(n)
// Space: O(1)
// dp
class Solution2 {
public:
    int countHousePlacements(int n) {
        static const int MOD = 1e9 + 7;
        int64_t prev = 0, curr = 1;
        for (int _ = 0; _ < n + 1; ++_) {
            tie(prev, curr) = pair(curr, (prev + curr) % MOD);
        }
        return curr * curr % MOD;
    }
};
